app/scripts/metamask-controller.js

```diff
@@ -4885,52 +4892,76 @@ export default class MetamaskController extends EventEmitter {
   // Hardware
   //
 
-  async getKeyringForDevice(deviceName, hdPath = null) {
+  /**
+   * Select a hardware wallet device and execute a
+   * callback with the keyring for that device.
+   *
+   * Note that KeyringController state is not updated before
+   * the end of the callback execution, and calls to KeyringController
+   * methods within the callback can lead to deadlocks.
+   *
+   * @param {object} options - The options for the device
+   * @param {string} options.name - The device name to select
+   * @param {string} options.hdPath - An optional hd path to be set on the device
+   * keyring
+   * @param {*} callback - The callback to execute with the keyring
+   * @returns {*} The result of the callback
+   */
+  async withKeyringForDevice(options, callback) {
```

Comment:
Nit: can we make this private I see it's being used only on this file (can we have a workaround for the test file without mocking a private function)?

Reply:
There's no other component relying on this method besides MetamaskController itself so I believe that it can be indeed made private (#withKeyringForDevice), though as you are suggesting there are a couple of test cases where it's being mocked. We could try to mock `KeyringController.withKeyring` instead, since it's the function that withKeyringForDevice relies upon

Reply:
Done in https://github.com/MetaMask/metamask-extension/pull/27025/commits/bfda15f6caffa7648eb2ff32a735c8df8fc4f500